contracts/lib/MerklePatriciaProof.sol

```diff
@@ -60,6 +60,7 @@ library MerklePatriciaProof {
                 nodeKey = RLP.toBytes32(currentNodeList[nextPathNibble]);
                 pathPtr += 1;
             } else if(currentNodeList.length == 2) {
+                uint currentpathPtr = pathPtr;
                 pathPtr += _nibblesToTraverse(RLP.toData(currentNodeList[0]), path, pathPtr);
```

Comment:
we potentially calculate `_nibblesToTraverse(RLP.toData(currentNodeList[0]), path, pathPtr)` twice.

suggestion:
```
uint traverseLength = _nibblesToTraverse(RLP.toData(currentNodeList[0]), path, pathPtr);
if(pathPtr + traverseLength == path.length) { //leaf node
            if(keccak256(abi.encodePacked(RLP.toData(currentNodeList[1]))) == value) {
                   return true;
            } else {
                   return false;
            }
} else if (traverseLength == 0) { // error: couldn't traverse path
           return false
} else { // extension node
           pathPtr += traverseLength;
}

Reply:
The last `else` block could also include the `nodeKey = RLP.toBytes32(currentNodeList[1]);` part.

Reply:
> we potentially calculate `_nibblesToTraverse(RLP.toData(currentNodeList[0]), path, pathPtr)` twice.
> 
> suggestion:
> 
> ```
> uint traverseLength = _nibblesToTraverse(RLP.toData(currentNodeList[0]), path, pathPtr);
> if(pathPtr + traverseLength == path.length) { //leaf node
>             if(keccak256(abi.encodePacked(RLP.toData(currentNodeList[1]))) == value) {
>                    return true;
>             } else {
>                    return false;
>             }
> } else if (traverseLength == 0) { // error: couldn't traverse path
>            return false
> } else { // extension node
>            pathPtr += traverseLength;
> }
> ```

Are you fine with this? In the original code, pathPtr was always updated, Can we also keep like that? 
Though, your suggestion looks absolutely correct as per the specification. 
```
uint traverseLength = _nibblesToTraverse(RLP.toData(currentNodeList[0]), path, pathPtr);
if(pathPtr + traverseLength == path.length) { //leaf node
            if(keccak256(abi.encodePacked(RLP.toData(currentNodeList[1]))) == value) {
                   return true;
            } else {
                   return false;
            }
} else if (traverseLength == 0) { // error: couldn't traverse path
           return false
} 
 pathPtr += traverseLength;

```

Reply:
I think Ben's addition of the explicit `else` improves readability.